### PR TITLE
Syntax keywords support

### PR DESCRIPTION
**Pull request by ptrv**

### CODE REVIEW

**Comment by ptrv** [comment]
We could instead of loading keywords from file like it is in the PR, specify an alist with mode and keyword. The `defvar` could be in another file (i.e. `ycmd-keywords.el`)

```lisp
(defvar ycmd-keywords-alist
  '(c++-mode
    "keyword1" "keyword2" ... ))

(defun ycmd--get-keywords-from-alist (mode)
  "Get completions from `ycmd-keywords-alist' for MODE."
  (let ((completion-ignore-case nil)
        (symbols (cdr (assq mode ycmd-keywords-alist))))
    (if (consp symbols)
        symbols
      (cdr (assq symbols ycmd-keywords-alist)))))
```

@abingham What do you think? What should we use?

File: ycmd.el [hunk 1]
```diff
@@ -701,6 +721,42 @@ Returns the new value of `ycmd-force-semantic-completion'."
       (setq it (list it)))
     (mapcar 'expand-file-name it)))
 
+(defvar ycmd--mode-keywords-loaded nil)
+
+(defun ycmd--get-keywords (buffer)
+  "Get syntax keywords for BUFFER."
+  (with-current-buffer buffer
+    (let ((mode major-mode))
+      (unless (memq mode ycmd--mode-keywords-loaded)
```
Comment (abingham) [inline ycmd.el]: Won't this function return nil every time it's called after the first time for a mode? We populate ycmd--mode-keywords-loaded the first time, so subsequent calls will do nothing. What am I missing?

**Comment by abingham** [comment]
I'm definitely a bit hesitant about the current design. I feel like core of the feature should be the kind of alist you describe, and users can populate that how they see fit. We could provide some handy functions for populating that alist from files (because that's very convenient), but the choice to do so should be a bit more explicit than it is now, I think. 

One use-case I want to make sure we consider is someone who's experimenting with a new language and doesn't have a syntax-file for it. They should be able to e.g. populate the syntax structure from the scratch buffer while the rest of the system keeps working.

File: ycmd.el (hunk 1 above)
Comment (ptrv) [inline ycmd.el]: We only need to send the keywords once to the server, therefore the check. It's the same in the vim YCM plugin.

**Comment by ptrv** [comment]
After I started the PR I realized that I am also more in favor of having an alist. That's why I asked :smile: 

> We could provide some handy functions for populating that alist from files (because that's very convenient), but the choice to do so should be a bit more explicit than it is now, I think.

What do you mean by more explicit? I think the two file loading funcitons are pretty clear? Maybe I am missing something?

I would remove the file loading functions  and add the one which reads out the alist. There is still the possibility to set `ycmd-get-keywords-function` to a custom function which could load keywords from files or from other custom sources.

Should we keep the alist in `ycmd.el` or should we add a new file just for the alist?

**Comment by abingham** [comment]
My more explicit, I guess I mean that users should have to specifically request (e.g. through the emacs configuration) that they want to load syntax info from files. It shouldn't be the default behavior. The functions you've got are just fine. 

I'm fine with keeping the alist in ycmd.el, but if you feel that it's becoming unwieldy then go ahead and move it into another file. If you do that, we'll need to update the melpa recipe, but that's a simple procedure.

**Comment by ptrv** [comment]
PR updated.

Read keywords from an alist and remove code for loading keywords from files.

**Comment by abingham** [comment]
Looks good. I'm excited to try this with elm when I get home!

**Comment by ptrv** [comment]
Great! Thanks for merging

**Pull request merged**